File: python/src/spring_nbt_library/world/paletted_container.py
"""パレットとビットストレージの組
セクション内のブロック状態やバイオームを格納する

パレットの要素は**生の NbtTag のまま**持つ
こうすると、触っていないブロックについては Minecraft が書き出したときの
プロパティの並び順まで含めてそのまま書き戻せる

仕様: ``docs/spec/31-paletted-container.md``
"""

from __future__ import annotations

from typing import List, Optional, Sequence

from ..errors import SpringNbtError
from ..nbt import NbtCompound, NbtList, NbtLongArray, NbtTag
from .bit_storage import BitStorage

__all__ = ["PalettedContainer", "ceil_log2"]


def ceil_log2(count: int) -> int:
    """``count`` 個の値を表すのに必要な最小ビット数
    1 なら 0
    """
    bits = 0

    # 1 を超える分だけシフトして数える
    while (1 << bits) < count:
        bits += 1

    return bits


class PalettedContainer:
    """パレット付きのコンテナ"""

    __slots__ = ("_palette", "entry_count", "min_bits", "_storage")

    def __init__(self, entry_count: int, min_bits: int) -> None:
        self._palette: List[NbtTag] = []
        self.entry_count = entry_count
        self.min_bits = min_bits
        self._storage: Optional[BitStorage] = None

    @property
    def palette(self) -> Sequence[NbtTag]:
        """パレット
        読み取り専用として扱うこと
        """
        return self._palette

    @property
    def bits_per_entry(self) -> int:
        """現在のビット幅
        パレットが 1 要素なら 0（記憶域を持たない）
        """
        if self._storage is None:
            return 0

        return self._storage.bits_per_entry

    @staticmethod
    def filled(value: NbtTag, entry_count: int, min_bits: int) -> "PalettedContainer":
        """単一の値で埋めたコンテナを作る"""
        result = PalettedContainer(entry_count, min_bits)
        result._palette.append(value)
        return result

    @staticmethod
    def from_nbt(nbt: NbtCompound, entry_count: int, min_bits: int,
                 lenient_bit_storage: bool = False) -> "PalettedContainer":
        """NBT から読み込む

        :raises SpringNbtError: パレットが空、data の長さが合わない、添字が範囲外のいずれか
        """
        result = PalettedContainer(entry_count, min_bits)
        palette_tag = nbt.opt_list("palette")

        if palette_tag is None or len(palette_tag) == 0:
            raise SpringNbtError.malformed("palette が無いか空")

        # パレットの要素は生の NbtTag のまま持つ
        # 並び順まで元どおりに書き戻すため
        for entry in palette_tag:
            result._palette.append(entry)

        data = nbt.opt_long_array("data")

        if data is None:
            # パレットが 1 要素なら data は無くてよい
            if len(result._palette) != 1:
                raise SpringNbtError.malformed(
                    "palette が %d 要素なのに data が無い" % len(result._palette))

            return result

        bits = max(min_bits, ceil_log2(len(result._palette)))
        result._storage = BitStorage.from_longs(data, bits, entry_count, lenient_bit_storage)

        # 取り出した添字がパレットの範囲に収まっているか確かめる
        # 黙って 0 番目で代替すると、壊れたデータをそうと分からない形で書き戻してしまう
        for index in range(entry_count):
            value = result._storage.get(index)

            if value >= len(result._palette):
                raise SpringNbtError.malformed(
                    "添字 %d の値 %d がパレット（%d 要素）の範囲外"
                    % (index, value, len(result._palette)))

        return result

    def to_nbt(self) -> NbtCompound:
        """NBT へ変換する"""
        result = NbtCompound()
        palette_tag = NbtList()

        # パレットの要素は読んだときのまま書き出す
        for entry in self._palette:
            palette_tag.append(entry)

        # パレットが 1 要素なら data は書かない
        # Minecraft と同じ振る舞い
        if self._storage is not None and len(self._palette) > 1:
            result.set("data", NbtLongArray(self._storage.to_longs()))

        result.set("palette", palette_tag)
        return result

    def get(self, index: int) -> NbtTag:
        """添字の値を取り出す"""
        self._check_index(index)

        # 記憶域が無いということは、全エントリがパレットの 0 番目
        if self._storage is None:
            return self._palette[0]

        return self._palette[self._storage.get(index)]

    def set(self, index: int, value: NbtTag) -> None:
        """添字の値を書き換える
        パレットに無ければ追加する
        """
        self._check_index(index)
        palette_index = self._index_of_or_add(value)

        # 記憶域が無く、書き込む値も 0 番目なら何もしなくてよい
        if self._storage is None and palette_index == 0:
            return

        self._ensure_storage()
        self._storage.set(index, palette_index)

    def fill(self, value: NbtTag) -> None:
        """全エントリを 1 つの値で埋める
        パレットもその 1 要素だけにする
        """
        self._palette = [value]
        self._storage = None

    def compact(self) -> None:
        """どのエントリからも参照されていないパレット要素を取り除き、添字を振り直す

        大量の ``set`` を行う用途で遅くならないよう、明示的に呼んだときだけ実行する
        """
        if self._storage is None:
            return

        used_entries = [False] * len(self._palette)

        # どのパレット要素が実際に使われているかを数える
        for index in range(self.entry_count):
            used_entries[self._storage.get(index)] = True

        compacted = []
        remap = [-1] * len(self._palette)

        # 使われている要素だけを詰め直し、新しい添字を割り当てる
        for old in range(len(self._palette)):
            if not used_entries[old]:
                continue

            remap[old] = len(compacted)
            compacted.append(self._palette[old])

        if len(compacted) == len(self._palette):
            return

        new_bits = max(self.min_bits, ceil_log2(len(compacted)))
        rebuilt = BitStorage.create(new_bits, self.entry_count)

        # 新しい添字へ置き換えながら詰め直す
        for index in range(self.entry_count):
            rebuilt.set(index, remap[self._storage.get(index)])

        self._palette = compacted

        if len(compacted) == 1:
            # 1 要素になったら記憶域を捨てる
            self._storage = None
        else:
            self._storage = rebuilt

    def _index_of_or_add(self, value: NbtTag) -> int:
        """パレット内の位置を返す
        無ければ末尾へ追加する
        """
        # パレットは高々 4096 要素なので線形探索で足りる
        for index in range(len(self._palette)):
            if self._palette[index] == value:
                return index

        self._palette.append(value)
        return len(self._palette) - 1

    def _ensure_storage(self) -> None:
        """現在のパレット長に合うビット幅の記憶域を用意する"""
        required = max(self.min_bits, ceil_log2(len(self._palette)))

        if self._storage is None:
            # これまで単一値だったので、全エントリが 0 番目のまま始まる
            self._storage = BitStorage.create(required, self.entry_count)
            return

        if self._storage.bits_per_entry >= required:
            return

        # パレットが増えてビット幅が足りなくなったら、全体を詰め直す
        self._storage = self._storage.resize(required)

    def _check_index(self, index: int) -> None:
        if index < 0 or index >= self.entry_count:
            raise SpringNbtError.invalid_argument(
                "添字が範囲外: %d (0..%d)" % (index, self.entry_count - 1))

File: java/src/main/java/io/github/scriptarts/springnbt/world/PalettedContainer.java
package io.github.scriptarts.springnbt.world;

import io.github.scriptarts.springnbt.SpringNbtException;
import io.github.scriptarts.springnbt.nbt.NbtCompound;
import io.github.scriptarts.springnbt.nbt.NbtList;
import io.github.scriptarts.springnbt.nbt.NbtLongArray;
import io.github.scriptarts.springnbt.nbt.NbtTag;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/**
 * パレットとビットストレージの組。セクション内のブロック状態やバイオームを格納する。
 *
 * <p>パレットの要素は<strong>生の {@link NbtTag} のまま</strong>持つ。
 * こうすると、触っていないブロックについては Minecraft が書き出したときの
 * プロパティの並び順まで含めてそのまま書き戻せる。
 *
 * <p>仕様: {@code docs/spec/31-paletted-container.md}
 */
public final class PalettedContainer {

    private final List<NbtTag> palette = new ArrayList<>();
    private final int entryCount;
    private final int minBits;

    private BitStorage storage;

    private PalettedContainer(int entryCount, int minBits) {
        this.entryCount = entryCount;
        this.minBits = minBits;
    }

    /**
     * エントリ数。ブロックなら 4096、バイオームなら 64。
     *
     * @return エントリ数
     */
    public int entryCount() {
        return entryCount;
    }

    /**
     * ビット幅の下限。ブロックなら 4、バイオームなら 1。
     *
     * @return 下限
     */
    public int minBits() {
        return minBits;
    }

    /**
     * パレット。読み取り専用。
     *
     * @return パレット
     */
    public List<NbtTag> palette() {
        return Collections.unmodifiableList(palette);
    }

    /**
     * 現在のビット幅。パレットが 1 要素なら 0（記憶域を持たない）。
     *
     * @return ビット幅
     */
    public int bitsPerEntry() {
        if (storage == null) {
            return 0;
        }

        return storage.bitsPerEntry();
    }

    /**
     * 単一の値で埋めたコンテナを作る。
     *
     * @param value      値
     * @param entryCount エントリ数
     * @param minBits    ビット幅の下限
     * @return コンテナ
     */
    public static PalettedContainer filled(NbtTag value, int entryCount, int minBits) {
        Objects.requireNonNull(value, "value");
        PalettedContainer result = new PalettedContainer(entryCount, minBits);
        result.palette.add(value);
        return result;
    }

    /**
     * NBT から読み込む。
     *
     * @param nbt                 コンテナの NBT
     * @param entryCount          エントリ数
     * @param minBits             ビット幅の下限
     * @param lenientBitStorage   data の長さが合わないとき、長さから逆算するか
     * @return コンテナ
     * @throws SpringNbtException パレットが空、data の長さが合わない、添字が範囲外のいずれか
     */
    public static PalettedContainer fromNbt(
            NbtCompound nbt, int entryCount, int minBits, boolean lenientBitStorage) {
        Objects.requireNonNull(nbt, "nbt");
        PalettedContainer result = new PalettedContainer(entryCount, minBits);
        NbtList paletteTag = nbt.optList("palette");

        if (paletteTag == null || paletteTag.size() == 0) {
            throw SpringNbtException.malformed("palette が無いか空");
        }

        // パレットの要素は生の NbtTag のまま持つ。並び順まで元どおりに書き戻すため
        for (NbtTag entry : paletteTag) {
            result.palette.add(entry);
        }

        long[] data = nbt.optLongArray("data");

        if (data == null) {
            // パレットが 1 要素なら data は無くてよい
            if (result.palette.size() != 1) {
                throw SpringNbtException.malformed(
                        "palette が " + result.palette.size() + " 要素なのに data が無い");
            }

            return result;
        }

        int bits = Math.max(minBits, ceilLog2(result.palette.size()));
        result.storage = BitStorage.fromLongs(data, bits, entryCount, lenientBitStorage);

        // 取り出した添字がパレットの範囲に収まっているか確かめる。
        // 黙って 0 番目で代替すると、壊れたデータをそうと分からない形で書き戻してしまう
        for (int index = 0; index < entryCount; index++) {
            int value = result.storage.get(index);

            if (value >= result.palette.size()) {
                throw SpringNbtException.malformed("添字 " + index + " の値 " + value
                        + " がパレット（" + result.palette.size() + " 要素）の範囲外");
            }
        }

        return result;
    }

    /**
     * NBT へ変換する。
     *
     * @return NBT
     */
    public NbtCompound toNbt() {
        NbtCompound result = new NbtCompound();
        NbtList paletteTag = new NbtList();

        // パレットの要素は読んだときのまま書き出す
        for (NbtTag entry : palette) {
            paletteTag.add(entry);
        }

        // パレットが 1 要素なら data は書かない。Minecraft と同じ振る舞い
        if (storage != null && palette.size() > 1) {
            result.set("data", new NbtLongArray(storage.toLongs()));
        }

        result.set("palette", paletteTag);
        return result;
    }

    /**
     * 添字の値を取り出す。
     *
     * @param index 添字
     * @return 値
     */
    public NbtTag get(int index) {
        checkIndex(index);

        // 記憶域が無いということは、全エントリがパレットの 0 番目
        if (storage == null) {
            return palette.get(0);
        }

        return palette.get(storage.get(index));
    }

    /**
     * 添字の値を書き換える。パレットに無ければ追加する。
     *
     * @param index 添字
     * @param value 値
     */
    public void set(int index, NbtTag value) {
        Objects.requireNonNull(value, "value");
        checkIndex(index);

        int paletteIndex = indexOfOrAdd(value);

        // 記憶域が無く、書き込む値も 0 番目なら何もしなくてよい
        if (storage == null && paletteIndex == 0) {
            return;
        }

        ensureStorage();
        storage.set(index, paletteIndex);
    }

    /**
     * 全エントリを 1 つの値で埋める。パレットもその 1 要素だけにする。
     *
     * @param value 値
     */
    public void fill(NbtTag value) {
        Objects.requireNonNull(value, "value");
        palette.clear();
        palette.add(value);
        storage = null;
    }

    /**
     * どのエントリからも参照されていないパレット要素を取り除き、添字を振り直す。
     *
     * <p>大量の {@code set} を行う用途で遅くならないよう、明示的に呼んだときだけ実行する。
     */
    public void compact() {
        if (storage == null) {
            return;
        }

        boolean[] usedEntries = new boolean[palette.size()];

        // どのパレット要素が実際に使われているかを数える
        for (int index = 0; index < entryCount; index++) {
            usedEntries[storage.get(index)] = true;
        }

        List<NbtTag> compacted = new ArrayList<>();
        int[] remap = new int[palette.size()];

        // 使われている要素だけを詰め直し、新しい添字を割り当てる
        for (int old = 0; old < palette.size(); old++) {
            if (!usedEntries[old]) {
                remap[old] = -1;
                continue;
            }

            remap[old] = compacted.size();
            compacted.add(palette.get(old));
        }

        if (compacted.size() == palette.size()) {
            return;
        }

        int newBits = Math.max(minBits, ceilLog2(compacted.size()));
        BitStorage rebuilt = BitStorage.create(newBits, entryCount);

        // 新しい添字へ置き換えながら詰め直す
        for (int index = 0; index < entryCount; index++) {
            rebuilt.set(index, remap[storage.get(index)]);
        }

        palette.clear();
        palette.addAll(compacted);

        if (compacted.size() == 1) {
            // 1 要素になったら記憶域を捨てる
            storage = null;
        } else {
            storage = rebuilt;
        }
    }

    /** パレット内の位置を返す。無ければ末尾へ追加する。 */
    private int indexOfOrAdd(NbtTag value) {
        // パレットは高々 4096 要素なので線形探索で足りる
        for (int index = 0; index < palette.size(); index++) {
            if (palette.get(index).equals(value)) {
                return index;
            }
        }

        palette.add(value);
        return palette.size() - 1;
    }

    /** 現在のパレット長に合うビット幅の記憶域を用意する。 */
    private void ensureStorage() {
        int required = Math.max(minBits, ceilLog2(palette.size()));

        if (storage == null) {
            // これまで単一値だったので、全エントリが 0 番目のまま始まる
            storage = BitStorage.create(required, entryCount);
            return;
        }

        if (storage.bitsPerEntry() >= required) {
            return;
        }

        // パレットが増えてビット幅が足りなくなったら、全体を詰め直す
        storage = storage.resize(required);
    }

    private void checkIndex(int index) {
        if (index < 0 || index >= entryCount) {
            throw SpringNbtException.invalidArgument(
                    "添字が範囲外: " + index + " (0.." + (entryCount - 1) + ")");
        }
    }

    /**
     * {@code count} 個の値を表すのに必要な最小ビット数。1 なら 0。
     *
     * @param count 値の個数
     * @return ビット数
     */
    public static int ceilLog2(int count) {
        int bits = 0;

        // 1 を超える分だけシフトして数える
        while ((1 << bits) < count) {
            bits += 1;
        }

        return bits;
    }
}
